[Kubernetes provider]Update kubernetes provider README (#5990)
elastic/elastic-agent: internal/pkg/composable/providers/kubernetes/README.md

## Autodiscover provider in Elastic Agent

  

https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html

  

Kubernetes provider can be configured both in **standalone** and **fleet managed** elastic agent.

a. In  **standalone** elastic agent, the kubernetes provider is configured inside the [elastic-agent-standalone-kubernetes.yaml](https://github.com/elastic/elastic-agent/blob/f994f5bfdf68db27902a4175c3b655b4d611cf7c/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L28). User can update the parameters based on their needs.

b. In **fleet managed** elastic agent, kubernetes provider is enabled by default with default values. User needs to follow the steps described in the [documentation](https://www.elastic.co/guide/en/fleet/current/advanced-kubernetes-managed-by-fleet.html) to configure it.

> Condition based autodiscover is supported both in **fleet managed** elastic agent and in **standalone** elastic agent(see [doc](https://www.elastic.co/guide/en/fleet/current/conditions-based-autodiscover.html)).

Reviewer: ```suggestion
Condition based autodiscover is supported both in **fleet managed** elastic agent and in **standalone** elastic agent (see [doc](https://www.elastic.co/guide/en/fleet/current/conditions-based-autodiscover.html)).
```

## Autodiscover provider in Elastic Agent

  

https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html

  

Kubernetes provider can be configured both in **standalone** and **fleet managed** elastic agent.

> a. In  **standalone** elastic agent, user needs to edit the [elastic-agent-standalone-kubernetes.yaml](https://github.com/elastic/elastic-agent/blob/f994f5bfdf68db27902a4175c3b655b4d611cf7c/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L28)

Reviewer 2: Isn't this provider enabled by default in general? I believe this is the case for all providers.

Author: You mean if the user removes at all the `kubernetes.provider` block? I don't know then. Maybe I should state it differently, as user does not need to edit the configuration. It is there(in the manifest) with default values.

Reviewer 2: I believe it's enabled by default if no configuration is provided, but your current edit is fine.

> b. In **fleet managed** elastic agent, kubernetes provider is enabled by default with default values. User needs to follow the steps described in the [documentation](https://www.elastic.co/guide/en/fleet/current/advanced-kubernetes-managed-by-fleet.html) to configure it.

Reviewer 3: I feel like a blank line is missing from here

Reviewer 4: Does that cause it to render different? I don't think so.

Reviewer 3: it actually does, now **a.** and **b.** are in the same paragraph which makes it kind hard to distinguish the two. However if you add an empty line between **a.** and **b.** they appear properly separated 🙂

Reviewer 4: Then absolutely fix before merge.

Author: A blank line is needed indeed

Condition based autodiscover is supported both in **fleet managed** elastic agent and in **standalone** elastic agent(see [doc](https://www.elastic.co/guide/en/fleet/current/conditions-based-autodiscover.html)).

Hints based autodiscover is only supported in **standalone** elastic agent(see [doc](https://www.elastic.co/guide/en/fleet/current/hints-annotations-autodiscovery.html)).

  

### Conditions based autodiscover

  

> Example:

Reviewer 3: with the new write up below Example: feels redundant

As an example we will use the redis module.

> To automatically identify a Redis Pod and monitor it with the Redis integration, add the following input configuration inside the [Elastic Agent Standalone manifest](https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml):

Reviewer 3: qq: for the Conditions-based autodiscover to work correctly is any kubernetes provider required to be enabled? If yes please provide the necessary settings of the provider in this example

Reviewer 4: The default settings of the kubernetes provider should be enough for conditions to work.

      condition: ${kubernetes.pod.labels.app} == 'redis'
```

  

What makes this input block dynamic are the variables under hosts block and the condition.

`${kubernetes.pod.ip}` and `${kubernetes.pod.labels.app} == 'redis'`

  

#### High level description

> The Kubernetes provider watches for Kubernetes resources and generates mappings from them (similar to events in beats provider). The mappings include those variables([list of variables](https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html#_provider_for_pod_resources)) for each k8s resource with unique value for each one of them.

Reviewer: ```suggestion
The Kubernetes provider watches for Kubernetes resources and generates mappings from them (similar to events in beats provider). The mappings include those variables ([list of variables](https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html#_provider_for_pod_resources)) for each k8s resource with unique value for each one of them.
```

The updated complete inputs block will be then forwarded to agent to spawn/update metricbeat and filebeat instances.

### Hints based autodiscover

Standalone elastic agent supports autodiscover based on hints collected from the [Kubernetes Provider](https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html). The hints mechanism looks for hints in kubernetes pod annotations that have the prefix `co.elastic.hints`. As soon as the Pod is ready, elastic agent checks it for hints and launches the proper configuration for the container. Hints tell elastic agent how to monitor the container by using the proper integration.
The full list of supported hints can be found [here](https://www.elastic.co/guide/en/fleet/current/hints-annotations-autodiscovery.html#_required_hints).

Example:

As an example we will use again redis module.
Add the following annotations to a redis pod. Elastic agent will then initiate a new input with redis module to properly monitor the redis pod.
```
apiVersion: v1 
kind: Pod
metadata:
  name: redis 
  annotations:
    co.elastic.hints/package: redis
    co.elastic.hints/data_streams: info
    co.elastic.hints/info.period: 5m
```

> In order to enable hints based autodiscover, user needs to uncomment the following lines in the [Elastic Agent Standalone manifest](https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml)

Reviewer 2: @pkoutsovasilis we should make a note to update this with Helm Chart specific instructions after it's released.

      condition: ${kubernetes.pod.labels.app} == 'redis'
```

  

What makes this input block dynamic are the variables under hosts block and the condition.

`${kubernetes.pod.ip}` and `${kubernetes.pod.labels.app} == 'redis'`

  

#### High level description

The Kubernetes provider watches for Kubernetes resources and generates mappings from them (similar to events in beats provider). The mappings include those variables([list of variables](https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html#_provider_for_pod_resources)) for each k8s resource with unique value for each one of them.

> Agent composable controller which controls all the providers receives these mappings and tries to match them with the input blogs of the configurations.

Reviewer 2: ```suggestion
Agent composable controller which controls all the providers receives these mappings and tries to match them with the input blocks of the configurations.
```

Also, maybe add a link to the controller module here?

This means that for every mapping that the condition matches (`kubernetes.pod.labels.app == redis`), a new input will be created in which the condition will be removed(not needed anymore) and the `kubernetes.pod.ip` variable will be substituted from the value in the same mapping.

The updated complete inputs block will be then forwarded to agent to spawn/update metricbeat and filebeat instances.

### Hints based autodiscover

> Standalone elastic agent supports autodiscover based on hints collected from the [Kubernetes Provider](https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html). The hints mechanism looks for hints in kubernetes pod annotations that have the prefix `co.elastic.hints`. As soon as the container starts, elastic agent checks it for hints and launches the proper configuration for the container. Hints tell elastic agent how to monitor the container by using the proper integration.

Reviewer 2: ```suggestion
Standalone elastic agent supports autodiscover based on hints collected from the [Kubernetes Provider](https://www.elastic.co/guide/en/fleet/current/kubernetes-provider.html). The hints mechanism looks for hints in kubernetes pod annotations that have the prefix `co.elastic.hints`. As soon as the Pod is ready, elastic agent checks it for hints and launches the proper configuration for the container. Hints tell elastic agent how to monitor the container by using the proper integration.
```

The full list of supported hints can be found [here](https://www.elastic.co/guide/en/fleet/current/hints-annotations-autodiscovery.html#_required_hints).

Example:

As an example we will use again redis module.
Add the following annotations to a redis pod. Elastic agent will then initiate a new input with redis module to properly monitor the redis pod.
```
apiVersion: v1 
kind: Pod
metadata:
  name: redis 
  annotations:
    co.elastic.hints/package: redis
    co.elastic.hints/data_streams: info
    co.elastic.hints/info.period: 5m
```

> In order to enable hints based autodiscover, user needs to uncomment the following lines in the [Elastic Agent Standalone manifest](https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml)

Reviewer 3: point to the lines in the k8s yaml (https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L31) or even better, if you can, capture which agent providers configuration is needed and then point to the lines

2. Add the [volumeMount](https://github.com/elastic/elastic-agent/blob/c01636e7383a9b2af9a588e0fcf1a4cae7d0d65c/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L783-L785) and [volume](https://github.com/elastic/elastic-agent/blob/c01636e7383a9b2af9a588e0fcf1a4cae7d0d65c/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L824-L826).

The init container will start before the elastic agent pod and will donwload all the templates of packages [supported](https://github.com/elastic/elastic-agent/tree/main/deploy/kubernetes/elastic-agent-standalone/templates.d). 
Elastic agent will then collect from the pods running(through the watchers mechanism) all the hints annotations and will try to match them with the correct package.
In the redis example, it will use the [redis template](https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/elastic-agent-standalone/templates.d/redis.yml) and substitute the template variables with the values specified in the annotations. Default values will be used for variables not provided.

> A new input will be then created for redis and reds data will start being collected.

Reviewer 3: ```suggestion
A new input will be then created for redis and redis-specific data will start being collected.
```
